ksm.go

```diff
@@ -0,0 +1,303 @@
+	if attr.file == nil {
+		err := attr.open()
+		if err != nil {
+			return "", err
+		}
+
+		defer func() {
+			_ = attr.close()
+		}()
+	}
+
+	oldValue, err := attr.read()
+	if err != nil {
+		return "", err
+	}
+
+	err = attr.write(value)
+	if err != nil {
+		return "", err
+	}
+
+	return oldValue, nil
+}
+
+func isKSMAvailable(ksmDir string) error {
+	info, err := os.Stat(ksmDir)
+	if err != nil || !info.IsDir() {
+		return fmt.Errorf("%s is not available", ksmDir)
+	}
+
+	return nil
+}
+
+func systemPages() (int64, error) {
+	// We're going to parse meminfo
+	f, err := os.Open("/proc/meminfo")
+	if err != nil {
+		return -1, err
+	}
+	defer f.Close()
+
+	scan := bufio.NewScanner(f)
+	for scan.Scan() {
+		line := scan.Text()
+
+		// We only care about MemTotal
+		if !strings.HasPrefix(line, "MemTotal:") {
+			continue
+		}
+
+		// Extract the before last (value) and last (unit) fields
+		fields := strings.Split(line, " ")
+		value := fields[len(fields)-2]
+		totalMemory, err := strconv.ParseInt(value, 10, 64)
+		if err != nil {
+			return -1, fmt.Errorf("Invalid integer")
+		}
+
+		// meminfo gives us kB
+		totalMemory *= 1024
+
+		// Fetch the system page size
+		pageSize := (int64)(os.Getpagesize())
+
+		nPages := totalMemory / pageSize
+		return nPages, nil
+	}
+
+	return 0, fmt.Errorf("Could not compute number of pages")
+}
+
+func pagesToScan(s ksmSetting) (string, error) {
+	nPages, err := systemPages()
+	if err != nil {
+		return "", err
+	}
+
+	pagesToScan := nPages / s.pagesPerScanFactor
+
+	return fmt.Sprintf("%v", pagesToScan), nil
+}
+
+var initialPagesToScan string
+var initialSleepInterval string
+var initialKSMRun string
+
+func ksmTune(s ksmSetting) error {
+	err := isKSMAvailable(ksmRoot)
+	if err != nil {
+		return err
+	}
+
+	pagesToScanSysFs := sysfsAttribute{
+		path: filepath.Join(ksmRoot, ksmPagesToScan),
+	}
+
+	sleepIntervalSysFs := sysfsAttribute{
+		path: filepath.Join(ksmRoot, ksmSleepMillisec),
+	}
+
+	runSysFs := sysfsAttribute{
+		path: filepath.Join(ksmRoot, ksmRunFile),
+	}
+
+	newPagesToScan, err := pagesToScan(s)
+	if err != nil {
+		return err
+	}
+
+	initialPagesToScan, err = pagesToScanSysFs.store(newPagesToScan)
+	if err != nil {
+		return err
+	}
+
+	initialSleepInterval, err = sleepIntervalSysFs.store(fmt.Sprintf("%v", s.scanIntervalMs))
+	if err != nil {
+		fmt.Printf("Could not set slee interval %s\n", err)
+		return err
+	}
+
+	initialKSMRun, err = runSysFs.store(ksmRun)
+	if err != nil {
+		fmt.Printf("Could not set run %v\n", err)
+		return err
+	}
+
+	return nil
+}
+
+func ksmRestore() {
+	fmt.Printf("Restoring\n")
+	pagesToScanSysFs := sysfsAttribute{
+		path: filepath.Join(ksmRoot, ksmPagesToScan),
+	}
+
+	sleepIntervalSysFs := sysfsAttribute{
+		path: filepath.Join(ksmRoot, ksmSleepMillisec),
+	}
+
+	runSysFs := sysfsAttribute{
+		path: filepath.Join(ksmRoot, ksmRunFile),
+	}
+
+	_, err := pagesToScanSysFs.store(initialPagesToScan)
```

Comment:
I think it would be worth logging the original values "just in case" (also applies to the other two `.store()` calls below).